HIP/hip-859.md

```diff
@@ -0,0 +1,158 @@
+
+In this way we lose track of the actual gas that was consumed by the EVM.
+
+This HIP will add a new field `gas_consumed` to the `contract_result` table in the mirror node, which will contain the actually consumed gas by the EVM.
+
+```
+create table if not exists contract_result
+(
+...
+gas_limit            bigint       not null,
+gas_consumed         bigint       null,
+gas_used             bigint       null,
+...
+)
+```
+
+## Motivation
+
+Adding a `gas_consumed` field in the `contract_result` table would enrich user experience by providing more information about the transaction. 
+This would allow users to better understand the gas refund policy and the actual gas consumption of their transactions. This would also allow users to understand how heavy their executed transaction was for the EVM.
+
+## Rationale
+
+Consensus nodes export ContractActions sidecar records for each transaction after the introduction of [HIP-513](./hip-513.md). Each sidecar record contains a `gasUsed` field for the given action that the record represents (different type of call or contract creation).
+This value directly represents the gas consumed by the EVM, without taking into account the gas refund policy. This makes it possible to calculate the overall gas consumption for the transaction, during the ContractActions sidecar record importing.  
+
+If we sum up all of the `gasUsed` values of each ContractActions sidecar for a given transaction on top of the initial intrinsic gas, which is the initial gas, charged prior to the transaction execution, then we would get accurate value for the `gasConsumed` field.
+
+## User stories
+
+1. As a user, I want to see how much gas was consumed by the EVM during the execution of my transaction, when I retrieve a transaction info from the mirror-node REST APIs, so that I know if I can optimise my transaction cost
+2. As a user, I want to see and compare how much gas I was charged for and how much gas was consumed by the EVM, when I inspect my transaction in hashscan
+
+## Specification
+
+In order to calculate the `gasUsed` field, the mirror node should have `CONTRACT_ACTION` and `CONTRACT_BYTECODE` sidecars enabled. Then, we would use an enhancement logic in the `mirror-node-importer`,
+which will iterate over all `ContractAction` sidecar records for the transaction we want to import and calculate the `gasConsumed` field for.
+
+The new logic would have a mechanism to first calculate the intrinsic gas cost for the transaction being imported. We can inspect if the transaction has a `ContractByteCode` sidecar
+related to it. If yes, we have a contract creation transaction and we would calculate the intrinsic gas cost based on the length of the init bytecode. If we don't have such a sidecar,
+then the transaction represents a contract call and we would use the default intrinsic gas cost for contract calls, which is 21 000.
```

Comment:
I have a concern about this hardcoded 21_000 intrinsic gas cost. Has it ever been something different in the past? Is it available in the record stream anywhere like 0.0.121? How would we keep this in sync with services and have different values over time?

Reply:
I don't think this value was ever changed in Hedera and it is described in the Ethereum yellow paper. That's the basic minimum cost we charge for any EVM operation. That's the cost of a basic transfer in Ethereum. I suspect if this value ever change it would be very rare and it shouldn't be an issue. Or we can just not point out the concrete value and just mention "intrinsic gas cost" in the HIP?

Reply:
I'm fine with 21,000 in the HIP if it's never changed and is standard in Ethereum.

Reply:
It's standard in Ethereum and all networks utilizing the EVM. 
Any change would require a new EIP and a new EVM version and likely HIP to support it.